diff --strip-trailing-cr -urN Original/Paper-API/src/main/java/org/bukkit/plugin/SimplePluginManager.java Paper-API/src/main/java/org/bukkit/plugin/SimplePluginManager.java
--- Original/Paper-API/src/main/java/org/bukkit/plugin/SimplePluginManager.java	2021-08-21 15:30:09.702265700 +0900
+++ Paper-API/src/main/java/org/bukkit/plugin/SimplePluginManager.java	2021-09-01 05:09:25.711245800 +0900
@@ -20,6 +20,7 @@
 import java.util.Map;
 import java.util.Set;
 import java.util.WeakHashMap;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.logging.Level;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
@@ -60,6 +61,8 @@
     private final Map<String, Map<Permissible, Boolean>> permSubs = new HashMap<String, Map<Permissible, Boolean>>();
     private final Map<Boolean, Map<Permissible, Boolean>> defSubs = new HashMap<Boolean, Map<Permissible, Boolean>>();
     private boolean useTimings = false;
+    
+    private final ReentrantLock EVENT_LOCK = new ReentrantLock(true); // Chiyogami
 
     public SimplePluginManager(@NotNull Server instance, @NotNull SimpleCommandMap commandMap) {
         server = instance;
@@ -588,45 +591,50 @@
      */
     @Override
     public void callEvent(@NotNull Event event) {
-        // Paper - replace callEvent by merging to below method
-        if (event.isAsynchronous() && server.isPrimaryThread()) {
-            throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
-        } else if (!event.isAsynchronous() && !server.isPrimaryThread() && !server.isStopping() ) {
-            throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
-        }
-
-        HandlerList handlers = event.getHandlers();
-        RegisteredListener[] listeners = handlers.getRegisteredListeners();
-
-        for (RegisteredListener registration : listeners) {
-            if (!registration.getPlugin().isEnabled()) {
-                continue;
-            }
-
-            try {
-                registration.callEvent(event);
-            } catch (AuthorNagException ex) {
-                Plugin plugin = registration.getPlugin();
-
-                if (plugin.isNaggable()) {
-                    plugin.setNaggable(false);
-
-                    server.getLogger().log(Level.SEVERE, String.format(
+        try {
+            EVENT_LOCK.lock();
+            // Paper - replace callEvent by merging to below method
+            if (event.isAsynchronous() && server.isPrimaryThread()) {
+                throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
+            } else if (!event.isAsynchronous() && !server.isPrimaryThread() && !server.isStopping()) {
+                throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
+            }
+    
+            HandlerList handlers = event.getHandlers();
+            RegisteredListener[] listeners = handlers.getRegisteredListeners();
+    
+            for (RegisteredListener registration : listeners) {
+                if (!registration.getPlugin().isEnabled()) {
+                    continue;
+                }
+        
+                try {
+                    registration.callEvent(event);
+                } catch (AuthorNagException ex) {
+                    Plugin plugin = registration.getPlugin();
+            
+                    if (plugin.isNaggable()) {
+                        plugin.setNaggable(false);
+                
+                        server.getLogger().log(Level.SEVERE, String.format(
                             "Nag author(s): '%s' of '%s' about the following: %s",
                             plugin.getDescription().getAuthors(),
                             plugin.getDescription().getFullName(),
                             ex.getMessage()
-                            ));
-                }
-            } catch (Throwable ex) {
-                // Paper start - error reporting
-                String msg = "Could not pass event " + event.getEventName() + " to " + registration.getPlugin().getDescription().getFullName();
-                server.getLogger().log(Level.SEVERE, msg, ex);
-                if (!(event instanceof ServerExceptionEvent)) { // We don't want to cause an endless event loop
-                    callEvent(new ServerExceptionEvent(new ServerEventException(msg, ex, registration.getPlugin(), registration.getListener(), event)));
+                        ));
+                    }
+                } catch (Throwable ex) {
+                    // Paper start - error reporting
+                    String msg = "Could not pass event " + event.getEventName() + " to " + registration.getPlugin().getDescription().getFullName();
+                    server.getLogger().log(Level.SEVERE, msg, ex);
+                    if (!(event instanceof ServerExceptionEvent)) { // We don't want to cause an endless event loop
+                        callEvent(new ServerExceptionEvent(new ServerEventException(msg, ex, registration.getPlugin(), registration.getListener(), event)));
+                    }
+                    // Paper end
                 }
-                // Paper end
             }
+        } finally {
+            EVENT_LOCK.unlock();
         }
     }
 
