From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bea4dev <bea0224@outlook.jp>
Date: Tue, 29 Apr 2025 01:33:00 +0900
Subject: [PATCH] add world threading


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index cc35c6271c98615a486422d0490145227626bd85..48f45a29d8cdca1d0e823ee1bafc974c5c28e0f8 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1837,6 +1837,12 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
         this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
         for (ServerLevel level : this.getAllLevels()) {
+            // Chiyogami start - world threading
+            world.chiyogami.thread.WorldThread worldThread = level.worldThread;
+            if (worldThread.getWorldTickRunnable() == null) {
+            worldThread.setWorldTickRunnable(() -> {
+            level.thread = Thread.currentThread();
+            // Chiyogami end
             level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
             level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
             level.updateLagCompensationTick(); // Paper - lag compensation
@@ -1855,7 +1861,9 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             profiler.pop();
             profiler.pop();
             level.explosionDensityCache.clear(); // Paper - Optimize explosions
+            });} // Chiyogami - world threading
         }
+        world.chiyogami.thread.WorldThreadPool.doTick(this.levels.values()); // Chiyogami - world threading
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
         profiler.popPush("connection");
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 56659181dc817ea028428eabf5937308c04b8cc5..394aa2dfa6b40aa4343777b3d1568d4f903e3895 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -101,6 +101,7 @@ import org.bukkit.craftbukkit.CraftWorld;
 import org.bukkit.craftbukkit.block.CraftBlockState;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.bukkit.entity.SpawnCategory;
+import world.chiyogami.thread.WorldThread;
 // CraftBukkit end
 
 public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel, ca.spottedleaf.moonrise.patches.chunk_system.world.ChunkSystemEntityGetter { // Paper - rewrite chunk system // Paper - optimise collisions
@@ -122,7 +123,10 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     protected final CollectingNeighborUpdater neighborUpdater;
     private final List<TickingBlockEntity> pendingBlockEntityTickers = Lists.newArrayList();
     private boolean tickingBlockEntities;
-    public final Thread thread;
+    // Chiyogami start - world threading
+    public Thread thread;
+    public WorldThread worldThread;
+    // Chiyogami end
     private final boolean isDebug;
     private int skyDarken;
     protected int randValue = RandomSource.createThreadLocalInstance().nextInt();
