From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bea4dev <bea0224@outlook.jp>
Date: Tue, 29 Apr 2025 15:09:48 +0900
Subject: [PATCH] thread safe teleport


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 48477efbd01bb1f8987d9a3ae195710e36b7294f..0f7c0e4648e3861c082d4a772f10a4f361d406bd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -25,6 +25,7 @@ import java.util.UUID;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.BiConsumer;
 import java.util.function.Predicate;
+import java.util.function.Supplier;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.BlockUtil;
@@ -139,6 +140,7 @@ import net.minecraft.world.scores.PlayerTeam;
 import net.minecraft.world.scores.ScoreHolder;
 import net.minecraft.world.scores.Team;
 import org.jetbrains.annotations.Contract;
+import world.chiyogami.thread.WorldTask;
 
 public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess, ScoreHolder, DataComponentGetter, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity { // Paper - rewrite chunk system // Paper - optimise entity tracker
     // CraftBukkit start
@@ -3771,6 +3773,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
     }
 
     private Entity teleportCrossDimension(ServerLevel level, TeleportTransition teleportTransition) {
+        Supplier<Entity> teleportTask = () -> { // Chiyogami - thread safe teleport
         List<Entity> passengers = this.getPassengers();
         List<Entity> list = new ArrayList<>(passengers.size());
         this.ejectPassengers();
@@ -3808,6 +3811,10 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             profilerFiller.pop();
             return entityx;
         }
+        // Chiyogami start - thread safe teleport
+        };
+        return this.level().worldThread.runWorldTaskThreadSafely(teleportTransition.newLevel().worldThread, new WorldTask<>(teleportTask), "NMS Entity teleport");
+        // Chiyogami end
     }
 
     private TeleportTransition calculatePassengerTransition(TeleportTransition teleportTransition, Entity entity) {
