From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bea4dev <bea0224@outlook.jp>
Date: Tue, 29 Apr 2025 17:03:13 +0900
Subject: [PATCH] fix startup error


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index d4c94ef519e30d89d1976fba5ed09c7b7cfffefb..fd3553bdc1c3cdbf6aa3dc00e0a4987f8eaa4fb8 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -70,7 +70,6 @@ public class Main {
     @DontObfuscate
     public static void main(final OptionSet optionSet) { // CraftBukkit - replaces main(String[] args)
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
-        world.chiyogami.thread.ChiyogamiServerThreadProvider.init(); // Chiyogami - add ChiyogamiLib
         SharedConstants.tryDetectVersion();
         /* CraftBukkit start - Replace everything
         OptionParser optionParser = new OptionParser();
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 9f202225d2b9c63d38c45b3a2eda43383f63603d..173491520a7b2dbdfb35a0e4361ec61182ded46e 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -154,6 +154,10 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         world.chiyogami.config.ChiyogamiConfig.load();
         world.chiyogami.thread.WorldThreadPool.init();
         // Chiyogami end
+        // Chiyogami start - thread api
+        new world.chiyogami.thread.ImplConcurrentTaskHandler();
+        new world.chiyogami.thread.ImplWorldThreadLockHandler();
+        // Chiyogami end
 
         LOGGER.info("Loading properties");
         DedicatedServerProperties properties = this.settings.getProperties();
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 5f3ee846f7b63efba0c47dce17356f5b9d026d28..1a4e51b947251e4c3836e4de1560c7f11ba1cd40 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -113,7 +113,7 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
     private boolean tickingBlockEntities;
     // Chiyogami start - world threading
     public Thread thread;
-    public WorldThread worldThread;
+    public final WorldThread worldThread = new WorldThread(this);
     // Chiyogami end
     private final boolean isDebug;
     private int skyDarken;
