From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bea4dev <bea0224@outlook.jp>
Date: Tue, 29 Apr 2025 17:03:13 +0900
Subject: [PATCH] fix startup error


diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index 7d6d53c8eb2c724bd265e8ad15b22d71e97e2a7b..7070d543e5e20e7ae092b4b7ce0b2d8dd85045cb 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -28,6 +28,7 @@ import java.util.logging.Level;
 
 public class PaperEventManager {
     public static boolean synchronizedEvent; // Chiyogami - config
+    private final java.util.concurrent.locks.ReentrantLock EVENT_LOCK = new world.chiyogami.thread.WorldThreadSafeLock(true); // Chiyogami - synced event
 
     private final Server server;
 
@@ -35,8 +36,22 @@ public class PaperEventManager {
         this.server = server;
     }
 
+    // Chiyogami start - synced event
     // SimplePluginManager
     public void callEvent(@NotNull Event event) {
+        boolean shouldSync = synchronizedEvent && !event.isAsynchronous();
+        if (shouldSync) {
+            EVENT_LOCK.lock();
+            try {
+                callEventInternal(event);
+            } finally {
+                EVENT_LOCK.unlock();
+            }
+        } else {
+            callEventInternal(event);
+        }
+    }
+    void callEventInternal(@NotNull Event event) {
         if (event.isAsynchronous() && this.server.isPrimaryThread()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 71f7a8386d616420a0e86233e721f55ab62aae4d..4d683c70ef883fa072d3721b15a30c6561ef495b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -417,6 +417,7 @@ public final class CraftServer implements Server {
         this.serverLinks = new CraftServerLinks(console);
 
         Bukkit.setServer(this);
+        world.chiyogami.thread.ChiyogamiServerThreadProvider.init(); // Chiyogami - add ChiyogamiLib
         // Paper start
         this.commandMap = new CraftCommandMap(this);
         this.pluginManager = new SimplePluginManager(this, commandMap);
