From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bea4dev <bea0224@outlook.jp>
Date: Mon, 12 Jun 2023 19:44:29 +0900
Subject: [PATCH] chiyogami-init-commit-thread-1


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index ef36d793ab77c7b7208f8f5994815599cff470d1..5a0ae81a470ad77bb19c636088c202cb6dd36a4c 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -1783,6 +1783,29 @@ public final class Bukkit {
     public static boolean isPrimaryThread() {
         return server.isPrimaryThread();
     }
+    // Chiyogami start - thread
+    /**
+     * Checks if the current thread is the main thread.
+     * <p>
+     * If the current thread is a world thread, it is treated as not being the main thread.
+     *
+     * @return true if the current thread is the main thread,
+     *     false otherwise
+     */
+    public static boolean isMainThread(){
+        return server.isMainThread();
+    }
+
+    /**
+     * Checks if the current thread is a world thread.
+     *
+     * @return true if the current thread is a world thread,
+     *     false otherwise
+     */
+    public static boolean isWorldThread(){
+        return server.isWorldThread();
+    }
+    // Chiyogami end
 
     // Paper start
     /**
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 5b225bbb128893d67251a96ab318035802a0cf76..7de635bcde26a55af9cd5ab6053497390edd5e53 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1509,7 +1509,25 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      *     false otherwise
      */
     boolean isPrimaryThread();
+    // Chiyogami start
+    /**
+     * Checks if the current thread is the main thread.
+     * <p>
+     * If the current thread is a world thread, it is treated as not being the main thread.
+     *
+     * @return true if the current thread is the main thread,
+     *     false otherwise
+     */
+    boolean isMainThread();
 
+    /**
+     * Checks if the current thread is a world thread.
+     *
+     * @return true if the current thread is a world thread,
+     *     false otherwise
+     */
+    boolean isWorldThread();
+    // Chiyogami end
     // Paper start
     /**
      * Gets the message that is displayed on the server list.
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 72f1576b8ce5b55b50f053f346ce42c52db4b568..fdc7b2270965014f2c320b16416dd9f0353073e9 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -3908,6 +3908,31 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      */
     void setSendViewDistance(int viewDistance);
     // Paper end - view distance api
+    // Chiyogami start - thread
+    /**
+     * Schedules a task to the thread executing this world.
+     *
+     * @param runnable Tasks in runnable.
+     */
+    default void scheduleTask(Runnable runnable){scheduleTask(runnable, 0, 0);}
+
+    /**
+     * Schedules a task to the thread executing this world.
+     *
+     * @param runnable Tasks in runnable.
+     * @param delay Delay in executing a task.
+     */
+    default void scheduleTask(Runnable runnable, long delay){scheduleTask(runnable, 0, delay);}
+
+    /**
+     * Schedules a task to the thread executing this world.
+     *
+     * @param runnable Tasks in runnable.
+     * @param delay Delay in executing a task.
+     * @param period The interval at which the task is to be executed.
+     */
+    void scheduleTask(Runnable runnable, long delay, long period);
+    // Chiyogami end
 
     // Spigot start
     public class Spigot {
